### PR TITLE
feat: support --add-profile in sparo checkout & refactor current checkout profile logic

### PR DESCRIPTION
**Pull request by EscapeB**
### Basic Checks

Have you run `rush change` for this change?

- [x] Yes
- [ ] No

If **No**, please run `rush change` before, this is necessary.

If adding a **new feature**, the PR's description includes:

- [x] Reason for adding this feature
- [x] How to use
- [x] A basic example

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

### Summary
This PR add a new option called `--add-profile` in sparo checkout command, which can add profile and checkout projects in added profiles to current workspace.

### Detail

1. Add new add-profile option
```bash
# usage
sparo checkout main --add-profile sparo-development
```

2. Add two sparo profiles in repo.

### CODE REVIEW

File: apps/sparo-lib/src/cli/commands/checkout.ts [hunk 1]
```diff
@@ -108,6 +115,10 @@ export class CheckoutCommand implements ICommand<ICheckoutCommandOptions> {
         targetProfileNames = profiles;
       }
 
+      if (addProfiles.length) {
+        targetProfileNames.push(...addProfiles);
```
Comment (chengcyber) [inline apps/sparo-lib/src/cli/commands/checkout.ts:119]: I have a strong feeling now that the "targetProfileNames" should be a Set type to deduplicate same profile name.

Comment (EscapeB) [reply apps/sparo-lib/src/cli/commands/checkout.ts]: Good suggestion, I have changed all profiles' type to be Set<string>

**Comment by EscapeB** [comment]
@chengcyber ready for review now.

File: apps/sparo-lib/src/cli/commands/checkout.ts [hunk 2]
```diff
@@ -158,22 +167,40 @@ export class CheckoutCommand implements ICommand<ICheckoutCommandOptions> {
...
     localState.reset();
-    if (targetProfileNames.length) {
-      for (const p of targetProfileNames) {
+
+    if (isNoProfile) {
+      // if no profile specified, purge to skeleton
+      await this._gitSparseCheckoutService.purgeAsync();
+    } else if (targetProfileNames.size) {
+      // TODO: policy #1: Can not sparse checkout with uncommitted changes in the cone.
+      for (const p of profiles) {
```
Comment (octogonz) [inline apps/sparo-lib/src/cli/commands/checkout.ts:176]: spell out `profile` instead of `p`

File: apps/sparo-lib/src/cli/commands/checkout.ts [hunk 3]
```diff
@@ -230,24 +267,35 @@ export class CheckoutCommand implements ICommand<ICheckoutCommandOptions> {
...
+     * --add-profile is defined as array type parameter
+     * @example --no-profile --add-profile foo -> throw error
+     * @example --profile bar --add-profile foo -> current profiles = bar + foo
+     * @example --add-profile foo -> current profiles = current profiles + foo
+     */
+    const addProfiles: Set<string> = new Set(addProfilesFromArg.filter((p) => typeof p === 'string'));
+
+    if (isNoProfile && (profiles.size || addProfiles.size)) {
+      throw new Error(
+        `"--no-profile" can not be specified at the same time with "--profile" or "--add-profile"`
```
Comment (octogonz) [inline apps/sparo-lib/src/cli/commands/checkout.ts:291]:
```suggestion
        `The "--no-profile" parameter cannot be combined with "--profile" or "--add-profile"`
```

File: apps/sparo-lib/src/services/GitSparseCheckoutService.ts [hunk 4]
```diff
@@ -226,20 +225,30 @@ ${availableProfiles.join(',')}
...
+          // re-apply the initial paths for setting up sparse repo state
+          this._prepareMonorepoSkeleton({ restore: checkoutAction === 'purge' });
+          break;
+        case 'add':
+        case 'set':
+          if (targetFolders.length === 0) {
+            terminal.writeDebugLine(`Skip sparse checkout regarding no target folders`);
+          } else {
+            terminal.writeLine(
+              `Run sparse checkout ${checkoutAction} for these folders: ${targetFolders.join(' ')}`
```
Comment (octogonz) [inline apps/sparo-lib/src/services/GitSparseCheckoutService.ts:240]:
```suggestion
              `Performing sparse checkout ${checkoutAction} for these folders: ${targetFolders.join(' ')}`
```

Comment (octogonz) [reply apps/sparo-lib/src/services/GitSparseCheckoutService.ts:240]: This is a potentially long list of folders.  It should be `writeDebugLine()` instead of `writeLine()`, right?

If it's important to show the folders, then we should make sure they are formatted to be readable (e.g. `join('\n  ')`)

File: common/changes/sparo/feat-support_add_profile_in_checkout_2024-02-23-08-25.json [hunk 5]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "sparo",
+      "comment": "feat: support --add-profile in sparo checkout",
```
Comment (octogonz) [inline common/changes/sparo/feat-support_add_profile_in_checkout_2024-02-23-08-25.json:5]:
```suggestion
      "comment": "Improve \"sparo checkout\" to support the --add-profile parameter",
```
Use friendly language instead of "nerd voice" for CHANGELOG.md text

https://rushjs.io/pages/best_practices/change_logs/

File: common/changes/sparo/feat-support_add_profile_in_checkout_2024-02-26-04-10.json [hunk 6]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "sparo",
+      "comment": "refactor sparo checkout profile logic",
```
Comment (octogonz) [inline common/changes/sparo/feat-support_add_profile_in_checkout_2024-02-26-04-10.json:5]: Rephrase this to describe what benefit it provides for the end user (e.g. `Improve efficiency of the \"sparo checkout\" command` or `Fix handling of some edge cases for \"sparo checkout\"`).  If there is no observable benefit for end users, then don't document the change.

**Review by octogonz — APPROVED** [review]

File: apps/sparo-lib/src/services/GitSparseCheckoutService.ts (hunk 4 above)
Comment (EscapeB) [reply apps/sparo-lib/src/services/GitSparseCheckoutService.ts:240]: I think error message should be shown to users. Let me change it with `'\n  '`
